### PR TITLE
Add `hearth-canvas`

### PR DESCRIPTION
**Pull request by marceline-cramer**
Adds "canvas" support: rectangular RGB textures that can be efficiently and dynamically updated by the guest.

### CODE REVIEW

File: plugins/canvas/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+    ///
+    /// Does nothing until [Self::update_ubo] is called.
+    pub fn set_position(&mut self, position: Position) {
+        self.position = position;
+    }
+
+    /// Updates this draw's uniform buffer on the GPU.
+    pub fn update_ubo(&self, queue: &Queue, vp: Mat4) {
+        let half_size = vec2(self.position.half_size.x, -self.position.half_size.y);
+        let scale = Mat4::from_scale(half_size.extend(1.0));
```
Comment (airidaceae) [inline plugins/canvas/src/lib.rs:145]: document why we do this extend.

File: plugins/canvas/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+            // pass the texture size and add padding
+            Vec4::new(self.width as f32, self.height as f32, 0.0, 0.0)
+        };
+
+        let ubo = CanvasUniform { mvp, texture_size };
+
+        queue.write_buffer(&self.ubo, 0, bytemuck::bytes_of(&ubo));
+    }
+
+    pub fn blit(&self, queue: &Queue, mut blit: Blit) {
```
Comment (airidaceae) [inline plugins/canvas/src/lib.rs:164]: Doc comment

File: plugins/canvas/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+            },
+            Extent3d {
+                width,
+                height,
+                depth_or_array_layers: 1,
+            },
+        );
+    }
+
+    fn create_texture(device: &Device, queue: &Queue, mut pixels: Pixels) -> Texture {
```
Comment (airidaceae) [inline plugins/canvas/src/lib.rs:208]: doc comment

File: plugins/canvas/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+                sample_count: 1,
+                dimension: TextureDimension::D2,
+                format: TextureFormat::Rgba8UnormSrgb,
+                usage: TextureUsages::TEXTURE_BINDING | TextureUsages::COPY_DST,
+            },
+            &pixels.data,
+        )
+    }
+
+    fn create_bind_group(
```
Comment (airidaceae) [inline plugins/canvas/src/lib.rs:233]: doc comment

File: plugins/canvas/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+}
+
+impl ServiceRunner for CanvasFactory {
+    const NAME: &'static str = "hearth.canvas.CanvasFactory";
+
+    fn get_process_metadata() -> hearth_runtime::process::ProcessMetadata {
+        let mut meta = cargo_process_metadata!();
+
+        meta.description =
+            Some("The native canvas factory service. Accepts FactoryRequest".to_string());
```
Comment (airidaceae) [inline plugins/canvas/src/lib.rs:574]: End this with a period

File: plugins/canvas/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+                // create the instance
+                let instance = CanvasInstance {
+                    id,
+                    ops_tx: self.ops_tx.clone(),
+                };
+
+                // initialize the instance's metadata
+                let mut meta = cargo_process_metadata!();
+                meta.name = Some("CanvasInstance".to_string());
+                meta.description = Some("An instance of a canvas".to_string());
```
Comment (airidaceae) [inline plugins/canvas/src/lib.rs:537]: end with a period

File: core/schema/src/canvas.rs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    /// The height of the buffer, in pixels.
+    pub height: u32,
+
+    /// The RGBA color data of the buffer.
+    ///
+    /// `width * height * 4` should match the length of `data`. Missing pixel
+    /// data will be initialized with `0xff` for all components. Excess data
+    /// is ignored.
+    #[serde_as(as = "Base64")]
+    pub data: Vec<u8>,
```
[the remark this answers is not in the document]
Comment (marceline-cramer) [reply core/schema/src/canvas.rs:39]: Nope! <3

**Review by airidaceae — APPROVED** [review]

**Pull request merged**